### ISSUE TITLE
LSE atomics not supported by binutils

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#247]
Trying to compile arm64 with `CONFIG_ARM64_LSE_ATOMICS=y`, you might see:

```
arch/arm64/Makefile:40: LSE atomics not supported by binutils
arch/arm64/Makefile:48: Detected assembler with broken .inst; disassembly will be unreliable
```

This is because support is detected via the macro `as-instr` which invokes `$(CC)` without `KBUILD_CFLAGS`, so Clang's integrated assembler gets invoked, which doesn't understand the assembler directive `.arch_extension lse`.

It's not clear to me why `$(CC)` is invoked when testing for assembler support, rather than `$(AS)`. `as-option` in scripts/Kbuild.include also suffers from this.

On first observation, it looks like it should be simple to change the macro `as-instr`, but `as-option` needs its call sites updated as well.  Luckily, there are only 11.

This was first reported by @mkaehlcke in a CrOS bug I'll try to find and link to.  I misunderstood the source of error in that bug report and will need to update it.

So the main fix is kernel side, but it may be worthwhile to file a bug against Clang to understand that directive (though clang adding support for LSE atomics may be a far future project).

**Comment by nickdesaulniers** [comment]
```diff
diff --git a/scripts/Kbuild.include b/scripts/Kbuild.include
index ca21a35fa244..2b9564a0c9de 100644
--- a/scripts/Kbuild.include
+++ b/scripts/Kbuild.include
@@ -106,7 +106,7 @@ as-option = $(call try-run,\
 # Usage: cflags-y += $(call as-instr,instr,option1,option2)
 
 as-instr = $(call try-run,\
-       printf "%b\n" "$(1)" | $(CC) $(KBUILD_AFLAGS) -c -x assembler -o "$$TMP" -,$(2),$(3))
+       printf "%b\n" "$(1)" | $(AS) $(KBUILD_AFLAGS) -o "$$TMP" -,$(2),$(3))
 
 # __cc-option
 # Usage: MY_CFLAGS += $(call __cc-option,$(CC),$(MY_CFLAGS),-march=winchip-c6,-march=i586)
```
is the patch I'm testing.

**Comment by nickdesaulniers** [comment]
`as-option` might be a mess, since there's a large mix of `-Wa,` args being added to `KBUILD_CFLAGS` and other flags being added to `KBUILD_AFLAGS`.  What a mess. :hankey:

**Comment by nickdesaulniers** [comment]
Odd, Makefile sets `-no-integrated-as` for `KBUILD_AFLAGS` for clang...  Probably don't need `cc-option` on those.

**Comment by nickdesaulniers** [comment]
So clang is being invoked with:
```sh
$ printf "%b\n" ".arch_extension lse" | clang -no-integrated-as -c -x assembler -o a.out -
{standard input}: Assembler messages:
{standard input}:1: Error: unknown pseudo-op: `.arch_extension'
clang-8: error: assembler command failed with exit code 1 (use -v to see invocation)
```
so it's validating the assembly early (known issue) even though it's not assembling it.

**Comment by nathanchance** [comment]
I don't see this in mainline but I could reproduce it up until v4.16.

The following commands show me this was fixed by @agners in commit 0f0e8de334c54c38818a4a5390a39aa09deff5bf:

```
git bisect start --term-new=fixed --term-old=unfixed
git bisect fixed v4.16
git bisect unfixed v4.15
```

Bisect log:
```
git bisect start '--term-new=fixed' '--term-old=unfixed'
# fixed: [0adb32858b0bddf4ada5f364a84ed60b196dbcda] Linux 4.16
git bisect fixed 0adb32858b0bddf4ada5f364a84ed60b196dbcda
# unfixed: [d8a5b80568a9cb66810e75b182018e9edb68e8ff] Linux 4.15
git bisect unfixed d8a5b80568a9cb66810e75b182018e9edb68e8ff
# unfixed: [c14376de3a1befa70d9811ca2872d47367b48767] printk: Wake klogd when passing console_lock owner
git bisect unfixed c14376de3a1befa70d9811ca2872d47367b48767
# unfixed: [2246edfaf88dc368e8671b04afd54412625df60a] Merge tag 'for-linus' of git://git.kernel.org/pub/scm/linux/kernel/git/rdma/rdma
git bisect unfixed 2246edfaf88dc368e8671b04afd54412625df60a
# unfixed: [dfe8db22372873d205c78a9fd5370b1b088a2b87] Merge tag 'drm-misc-fixes-2018-02-21' of git://anongit.freedesktop.org/drm/drm-misc into drm-fixes
git bisect unfixed dfe8db22372873d205c78a9fd5370b1b088a2b87
# unfixed: [4665c6b04651e96c1e2eb9129a30d6055040ff73] Merge tag 'linux-can-fixes-for-4.16-20180312' of ssh://gitolite.kernel.org/pub/scm/linux/kernel/git/mkl/linux-can
git bisect unfixed 4665c6b04651e96c1e2eb9129a30d6055040ff73
# unfixed: [3215b9d57a2c75c4305a3956ca303d7004485200] Merge tag 'clk-fixes-for-linus' of git://git.kernel.org/pub/scm/linux/kernel/git/clk/linux
git bisect unfixed 3215b9d57a2c75c4305a3956ca303d7004485200
# unfixed: [ae42cc62a9f07f1f6979054ed92606b9c30f4a2e] bonding: move dev_mc_sync after master_upper_dev_link in bond_enslave
git bisect unfixed ae42cc62a9f07f1f6979054ed92606b9c30f4a2e
# unfixed: [a2601d78b77aacc5dd790f488188f9556f4a9eb2] Merge tag 'powerpc-4.16-6' of git://git.kernel.org/pub/scm/linux/kernel/git/powerpc/linux
git bisect unfixed a2601d78b77aacc5dd790f488188f9556f4a9eb2
# unfixed: [e81b5e01c14add8395dfba7130f8829206bb507d] net: mvneta: fix enable of all initialized RXQs
git bisect unfixed e81b5e01c14add8395dfba7130f8829206bb507d
# unfixed: [d89b9f50291a214f9d670594c9fc35483b6d87a4] Merge tag 'for-linus' of git://git.kernel.org/pub/scm/linux/kernel/git/rdma/rdma
git bisect unfixed d89b9f50291a214f9d670594c9fc35483b6d87a4
# unfixed: [a44406ec3d31de773ce50e794900546f22c1cc7b] Merge git://git.kernel.org/pub/scm/linux/kernel/git/davem/net
git bisect unfixed a44406ec3d31de773ce50e794900546f22c1cc7b
# fixed: [93e04d4ad7c567f87388dd0338df7d8878ab65e9] Merge branch 'x86-pti-for-linus' of git://git.kernel.org/pub/scm/linux/kernel/git/tip/tip
git bisect fixed 93e04d4ad7c567f87388dd0338df7d8878ab65e9
# fixed: [8a6105c4818cdc4a5c6b8a99c98ab7cc43be1ba8] kbuild: rpm-pkg: Support GNU tar >= 1.29
git bisect fixed 8a6105c4818cdc4a5c6b8a99c98ab7cc43be1ba8
# fixed: [0f0e8de334c54c38818a4a5390a39aa09deff5bf] kbuild: set no-integrated-as before incl. arch Makefile
git bisect fixed 0f0e8de334c54c38818a4a5390a39aa09deff5bf
# unfixed: [825d487583089f9a33d31650c9c41f6474aab7fc] kbuild: make scripts/adjust_autoksyms.sh robust against timestamp races
git bisect unfixed 825d487583089f9a33d31650c9c41f6474aab7fc
# first fixed commit: [0f0e8de334c54c38818a4a5390a39aa09deff5bf] kbuild: set no-integrated-as before incl. arch Makefile
```

It picks clean to v4.14.78, if we want to send it to Greg for stable inclusion.

**Comment by nickdesaulniers** [comment]
ah, sorry, my local build of Clang on my personal laptop was broken.  Can't repro at work.
Closing as dupe of #25 .  Thanks for bisecting the fix.  I'll send the patch to stable.
